toolchain/check/context.cpp

```diff
@@ -100,14 +100,13 @@ auto Context::AddPlaceholderInst(SemIR::ParseNodeAndInst parse_node_and_inst)
 
 auto Context::AddConstant(SemIR::Inst inst, bool is_symbolic)
     -> SemIR::ConstantId {
-  // TODO: Deduplicate constants.
-  auto inst_id = sem_ir().insts().AddInNoBlock(
-      SemIR::ParseNodeAndInst::Untyped(Parse::NodeId::Invalid, inst));
-  constants().Add(inst_id);
-
+  auto [inst_id, added] = constants().GetOrAdd(inst);
   auto const_id = is_symbolic ? SemIR::ConstantId::ForSymbolicConstant(inst_id)
                               : SemIR::ConstantId::ForTemplateConstant(inst_id);
-  constant_values().Set(inst_id, const_id);
+  if (added) {
+    // TODO: Should `ConstantStore` do this for us?
```

Comment:
Mostly I think I was thinking of consistency: adding to `insts()` doesn't change `constant_values()`, so adding to `constants()` seems like maybe it shouldn't do so either. And symmetrically, `Check::Context::AddInst` is responsible for updating `constant_values` in most cases, so having `SemIR` do it in some cases seems inconsistent.

I'm not sure what the right balance here is. The `ConstantStore` is a higher-level layer that is already modifying `insts()` (and reading from `inst_blocks()`), so it doesn't seem all that unreasonable for it to modify `constant_values()` too, but at the same time that seems like something that `Check` ought to be doing in the current setup.

I'm happy for you to tell me what you'd prefer here. If `ConstantStore` does start updating `constant_values()` too, should we even have a `Context::AddConstant` function at all, given that it would be a simple one-liner?

Reply:
> If `ConstantStore` does start updating `constant_values()` too, should we even have a `Context::AddConstant` function at all, given that it would be a simple one-liner?

The TODO had seemed reasonable to me in terms of not needing to return the "success" info. And actually, removing `AddConstant` would be the best argument for moving it to `constants`: it reduces the API surface of `Context`.